tiledb/sm/query/readers/ordered_dim_label_reader.cc

```diff
@@ -0,0 +1,914 @@
+                f, tile_idx, dim_dom[0], tile_extent);
+            if (covered) {
+              covered_tiles[f].emplace(tile_idx);
+            } else if (total_mem_used + tile_size < memory_budget_) {
+              total_mem_used += tile_size;
+              result_tiles_[f].emplace(
+                  std::piecewise_construct,
+                  std::forward_as_tuple(tile_idx),
+                  std::forward_as_tuple(f, frag_tile_idx, array_schema_));
+            } else {
+              if (r == 0) {
+                throw OrderedDimLabelReaderStatusException(
+                    "Can't process a single range, increase memory budget");
+              }
+              return r;
+            }
+          }
+        }
+      }
+    }
+    max_range = r;
+  }
+
+  return max_range + 1;
+}
+
+template <typename LabelType>
+LabelType OrderedDimLabelReader::get_label_value(
+    const unsigned f, const uint64_t tile_idx, const uint64_t cell_idx) {
+  auto& rt = result_tiles_[f].at(tile_idx);
+  const auto label_data =
+      rt.tile_tuple(label_name_)->fixed_tile().template data_as<LabelType>();
+  return label_data[cell_idx];
+}
+
+template <>
+std::string_view OrderedDimLabelReader::get_label_value<std::string_view>(
+    const unsigned f, const uint64_t tile_idx, const uint64_t cell_idx) {
+  auto& rt = result_tiles_[f][tile_idx];
+  auto tile_tuple = rt.tile_tuple(label_name_);
+  auto offsets_data = tile_tuple->fixed_tile().template data_as<uint64_t>();
+  auto& var_tile = tile_tuple->var_tile();
+  auto offset = offsets_data[cell_idx];
+
+  auto size = static_cast<size_t>(cell_idx) == rt.cell_num() - 1 ?
+                  var_tile.size() - offset :
+                  offsets_data[cell_idx + 1] - offset;
+  return std::string_view(&var_tile.template data_as<char>()[offset], size);
+}
+
+template <typename IndexType, typename LabelType>
+LabelType OrderedDimLabelReader::get_value_at(
+    const IndexType& index,
+    const IndexType& domain_low,
+    const IndexType& tile_extent) {
+  // Start with the most recent fragment.
+  for (auto f = static_cast<unsigned>(fragment_metadata_.size() - 1);; f--) {
+    const IndexType* non_empty_domain =
+        static_cast<const IndexType*>(non_empty_domains_[f]);
+    // If the value is in the non empty domain for the fragment, get it.
+    if (index >= non_empty_domain[0] && index <= non_empty_domain[1]) {
+      // Get the tile index in the current fragment.
+      const auto tile_idx =
+          index_dim_->tile_idx(index, domain_low, tile_extent);
+
+      // Get the cell index in the current tile.
+      const auto cell_idx =
+          index - index_dim_->tile_coord_low(tile_idx, domain_low, tile_extent);
+
+      // Finally get the data.
+      return get_label_value<LabelType>(f, tile_idx, cell_idx);
+    }
+
+    // We should always find the value before the last fragment.
+    if (f == 0) {
+      throw OrderedDimLabelReaderStatusException("Couldn't find value");
+    }
```

Comment:
I believe this error can show up if the non-empty domain contains a region with no data. Is that worth mentioning either in the comment or the error message?

Example use `CPPOrderedDimLabelReaderFixedDoubleFx`: 
write_labels(1, 4, {1.0, 2.0, 3.0, 4.0});
write_labels(10, 11, {8.0, 9.0});
std::vector<double> ranges{5.0, 10.0};
auto index = read_labels(ranges);
CHECK(index.size() == 0));